### ISSUE TITLE
Rust Analyzer shows wrong warning SHA1 hasher finalize

### ISSUE THREAD

**Issue by SamTV12345** [rust-lang/rust-analyzer#15242]
**rust-analyzer version**: (eg. output of "rust-analyzer: Show RA Version" command, accessible in VSCode via <kbd>Ctrl/⌘</kbd>+<kbd>Shift</kbd>+<kbd>P</kbd>)

**rustc version**: (eg. output of `rustc -V`) 1.70

**relevant settings**: (eg. client settings, or environment variables like `CARGO`, `RUSTC`, `RUSTUP_HOME` or `CARGO_HOME`): %

In IntelliJ the Rust plugin in version 0.4.197.5401-231 shows a warning when using the SHA-1 crate.

```Rust
        let mut hasher = Sha1::new();

        hasher.update(non_hashed_string);

        let hashed_auth_key = format!("{:x}", hasher.finalize());
```

In this part of the code it complains that hasher.finalize() `CtOutput<CoreWrapper<Sha1Core>>` doesn't implement `LowerHex` (required by {:x}). The code compiles and it also works in my open source production environment.

**Comment by lowr** [comment]
> In IntelliJ the Rust plugin in version 0.4.197.5401-231

This is rust-analyzer's repository, not [intellij-rust](https://github.com/intellij-rust/intellij-rust)'s. Can you check if you're reporting rust-analyzer's issue or intellij-rust's issue?

**Comment by lnicola** [comment]
We don't show any errors, but we don't resolve `Sha1::new`, so there's still a bug related to this.

**Comment by SamTV12345** [comment]
Thanks for taking a look at it 😄

**Comment by newpavlov** [comment]
Is it possible to bisect this? I think it's a relatively recent regression.

**Comment by lnicola** [comment]
The bug in my comment above is not a recent regression, but it was less noticeable before.


The "no such associated item" in https://github.com/RustCrypto/hashes/issues/529 is a recently-added diagnostic that makes the bug more visible. We've just marked it as experimental so it will go away next week, but that doesn't fix the underlying problem.

**Comment by lnicola** [comment]
Both of these are now working:

![image](https://github.com/user-attachments/assets/1edb200c-df0a-48ea-bf49-f489fab3a04f)
